### PR TITLE
AD-2486 Show all active TNPs

### PR DESCRIPTION
**Pull request by benagain**
Display TNP values from all active Price Episodes, instead of just the "first" price episode found in the database.

image

### CODE REVIEW

File: src/SFA.DAS.IdentifyDataLocks.Domain/CollectionPeriodMappingExtensions.cs [hunk 1]
```diff
@@ -66,14 +61,21 @@ public static class CollectionPeriodMappingExtensions
...
 
-        private static string GetTnpValue(decimal? tnp) => 
-            tnp?.ToString("c") ?? "-";
+        private static List<AmountFromDate> GetTnpValue(
+            EarningEventModel earning,
+            Func<EarningEventPriceEpisodeModel, decimal> tnpSelector)
+        {
+            return earning.PriceEpisodes.Select(AmountSelector).ToList();
+
+            AmountFromDate AmountSelector(EarningEventPriceEpisodeModel pe) =>
```
Comment (DevenKShah) [inline src/SFA.DAS.IdentifyDataLocks.Domain/CollectionPeriodMappingExtensions.cs:76]: This threw me off a bit . Can we have the in-line function defined before it is invoked.

File: src/SFA.DAS.IdentifyDataLocks.IntegrationTests/LearnerWithStoppedApprenticeship/LearnerWithStoppedApprenticeship.cs [hunk 2]
```diff
@@ -17,4 +20,102 @@ public async Task RendersDataMismatches()
...
+            var parser = new AngleSharp.Html.Parser.HtmlParser();
+            var document = parser.ParseDocument(body);
+            var tnp1Row = document.QuerySelector("tr.tnp1");
+            tnp1Row.Should().NotBeNull();
+            tnp1Row.Children.Should().BeEquivalentTo(
+                new[]
+                {
+                    new { TextContent = "TNP 1" },
+                    new { TextContent = "n/a" },
+                    new { TextContent = "Date: 01/12/2019\n        Price: �24,872.00" },
```
Comment (DevenKShah) [inline src/SFA.DAS.IdentifyDataLocks.IntegrationTests/LearnerWithStoppedApprenticeship/LearnerWithStoppedApprenticeship.cs:47]: Well done on this, but the white space and new line characters make these tests seem brittle.

**Review by DevenKShah — APPROVED** [review]
I have not tested it but looks good. Just couple of observations.

File: src/SFA.DAS.IdentifyDataLocks.IntegrationTests/LearnerWithStoppedApprenticeship/LearnerWithStoppedApprenticeship.cs (hunk 2 above)
Comment (benagain) [reply src/SFA.DAS.IdentifyDataLocks.IntegrationTests/LearnerWithStoppedApprenticeship/LearnerWithStoppedApprenticeship.cs:47]: Yes, I agree.  It's caused by the indentation in the .cshtml file.  I haven't come up with a way to deal with it yet.

File: src/SFA.DAS.IdentifyDataLocks.Domain/CollectionPeriodMappingExtensions.cs (hunk 1 above)
Comment (benagain) [reply src/SFA.DAS.IdentifyDataLocks.Domain/CollectionPeriodMappingExtensions.cs:76]: In C# we generally don't require class-level methods to be declared above their points of use, I see this as the same.

My preference is to see use of functions before their definition, with the rationale that I normally care most about what's going on and less about how it's being achieved.  Until I do care of course, and then I go looking for the function definition.

**Pull request merged**